Fabric.Authorization.Persistence.SqlServer/Stores/SqlServerRoleStore.cs

```diff
@@ -0,0 +1,132 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using Fabric.Authorization.Domain.Exceptions;
+using Fabric.Authorization.Domain.Models;
+using Fabric.Authorization.Domain.Stores;
+using Fabric.Authorization.Persistence.SqlServer.Mappers;
+using Fabric.Authorization.Persistence.SqlServer.Services;
+using Microsoft.EntityFrameworkCore;
+
+namespace Fabric.Authorization.Persistence.SqlServer.Stores
+{
+    public class SqlServerRoleStore : IRoleStore
+    {
+        private readonly IAuthorizationDbContext _authorizationDbContext;
+
+        public SqlServerRoleStore(IAuthorizationDbContext authorizationDbContext)
+        {
+            _authorizationDbContext = authorizationDbContext;
+        }
+
+        public async Task<Role> Add(Role model)
+        {
+            var entity = model.ToEntity();
+
+            _authorizationDbContext.Roles.Add(entity);
+            await _authorizationDbContext.SaveChangesAsync();
+            return model;
+        }
+
+        public async Task<Role> Get(Guid id)
+        {
+            var role = await _authorizationDbContext.Roles
+                .SingleOrDefaultAsync(r =>
+                    r.RoleId == id
+                    && !r.IsDeleted);
+
+            if (role == null)
+            {
+                throw new NotFoundException<Role>($"Could not find {typeof(Role).Name} entity with ID {id}");
+            }
+
+            return role.ToModel();
+        }
+
+        public async Task<IEnumerable<Role>> GetAll()
+        {
+            var roles = await _authorizationDbContext.Roles
+                .Where(r => !r.IsDeleted)
+                .ToArrayAsync();
+
+            return roles.Select(p => p.ToModel());
+        }
+
+        public async Task Delete(Role model)
+        {
+            var role = await _authorizationDbContext.Roles
+                .Include(r => r.RolePermissions)
+                .Include(r => r.GroupRoles)
+                .SingleOrDefaultAsync(r =>
+                    r.RoleId == model.Id
+                    && !r.IsDeleted);
+
+            if (role == null)
+            {
+                throw new NotFoundException<Role>($"Could not find {typeof(Role).Name} entity with ID {model.Id}");
+            }
+
+            role.IsDeleted = true;
+
+            foreach (var rolePermission in role.RolePermissions)
+            {
+                rolePermission.IsDeleted = true;
+            }
+
+            foreach (var groupRole in role.GroupRoles)
+            {
+                groupRole.IsDeleted = true;
+            }
+
+            await _authorizationDbContext.SaveChangesAsync();
```

Comment:
Should we consider wrapping this in a try/catch so we can deal with concurrency exceptions? In the case of a concurrency exception, we may want to retry the update. Not sure if we want it at this level, or if we want to handle it in the dbcontext, but something to think about.

Reply:
True - the same would apply to the permission store where we loop to delete the related entities. I'll hold off on adding it here until we decide whether we want to add it in the dbcontext.